nixos/modules/services/web-apps/pds.nix

```diff
@@ -0,0 +1,232 @@
+      '';
+    };
+
+    environmentFiles = mkOption {
+      type = types.listOf types.path;
+      default = [ ];
+      description = ''
+        File to load environment variables from. Loaded variables override
+        values set in {option}`environment`.
+
+        Use it to set values of `PDS_JWT_SECRET`, `PDS_ADMIN_PASSWORD`,
+        and `PDS_PLC_ROTATION_KEY_K256_PRIVATE_KEY_HEX` secrets.
+        You can generate initial values with
+        ```
+        nix-build -A pds.passthru.generateSecrets
+        ./result/bin/generate-pds-secrets > secrets.env
+        ```
+      '';
+    };
+
+    pdsadmin = {
+      enable = mkOption {
+        type = types.bool;
+        default = cfg.enable;
+        defaultText = literalExpression "services.pds.enable";
+        description = "Add pdsadmin script to PATH";
+      };
+
+      useServiceEnvironment = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Inherit environment variables of pds systemd unit";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment = mkIf cfg.pdsadmin.enable {
+      systemPackages = [
+        (if cfg.pdsadmin.useServiceEnvironment then pdsadminWrapper else pkgs.pdsadmin)
+      ];
+    };
+
+    systemd.services.pds = {
+      description = "pds";
+      wants = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = getExe cfg.package;
+        Environment = lib.mapAttrsToList (k: v: "${k}=${v}") (
+          lib.filterAttrs (_: v: v != null) cfg.settings
+        );
+
+        EnvironmentFile = cfg.environmentFiles;
+        User = "pds";
+        Group = "pds";
+        StateDirectory = "pds";
+        StateDirectoryMode = "0755";
+        Restart = "always";
+
+        # Hardening
+        RemoveIPC = true;
+        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" ];
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        ProtectClock = true;
+        ProtectKernelLogs = true;
+        ProtectControlGroups = true;
+        ProtectKernelModules = true;
+        PrivateMounts = true;
+        SystemCallArchitectures = [ "native" ];
+        MemoryDenyWriteExecute = false; # V8 JIT
+        RestrictNamespaces = true;
+        RestrictSUIDSGID = true;
+        ProtectHostname = true;
+        LockPersonality = true;
+        ProtectKernelTunables = true;
+        RestrictAddressFamilies = [
+          "AF_UNIX"
+          "AF_INET"
+          "AF_INET6"
+        ];
+        RestrictRealtime = true;
+        DeviceAllow = [ "" ];
+        ProtectSystem = "strict";
+        ProtectProc = "invisible";
+        ProcSubset = "pid";
+        ProtectHome = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        UMask = "0077";
+      };
+    };
+
+    users = {
+      users.pds = {
+        group = "pds";
+        createHome = false;
+        isSystemUser = true;
+      };
+      groups.pds = { };
+    };
```

Comment:
Is systemd `DynamicUser` not sufficient?

Reply:
AFAIK after reading the systemd.exec manpage, that wouldn't work since the StateDirectory stores state. Am I incorrect there?